src/core/shutdown.c

```diff
@@ -159,6 +164,102 @@ static int switch_root_initramfs(void) {
         return switch_root("/run/initramfs", "/oldroot", false, MS_BIND);
 }
 
+/* Read the following fields from /proc/meminfo:
+ *
+ *  NFS_Unstable
+ *  Writeback
+ *  Dirty
+ *
+ * Return true if the sum of these fields is greater than the previous
+ * value input. For all other issues, report the failure and indicate that
+ * the sync is not making progress.
+ */
+static bool sync_making_progress(unsigned long long *prev_dirty) {
+        _cleanup_fclose_ FILE *f = NULL;
+        char line[LINE_MAX];
+        int r = false;
+        unsigned long long val = 0;
+
+        f = fopen("/proc/meminfo", "re");
+        if (!f) {
+                return log_warning_errno(errno, "Failed to open /proc/meminfo: %m");
+        }
+
+        FOREACH_LINE(line, f, goto fail) {
+                unsigned long long ull = 0;
+
+                if (first_word(line, "NFS_Unstable:") || first_word(line, "Writeback:") || first_word(line, "Dirty:")) {
+                        if (sscanf(line, "%*s %llu %*s", &ull) != 1)
+                                goto fail;
+                } else
+                        continue;
+
+                val += ull;
+        }
+
+        r = *prev_dirty > val;
+
+        *prev_dirty = val;
+
+        return r;
+
+fail:
+        log_warning_errno(errno, "Failed to parse /proc/meminfo: %m");
+
+        return r;
+}
+
+static void sync_with_progress(void) {
+        unsigned checks;
+        pid_t pid;
+        int r;
+        unsigned long long dirty = ULONG_LONG_MAX;
+
+        BLOCK_SIGNALS(SIGCHLD);
+
+        /* Due to the possiblity of the sync operation hanging, we fork
+         * a child process and monitor the progress. If the timeout
+         * lapses, the assumption is that that particular sync stalled. */
+        pid = fork();
+        if (pid < 0) {
+                log_error_errno(errno, "Failed to fork: %m");
+                return;
+        }
+
+        if (pid == 0) {
+                /* Start the sync operation here in the child */
+                sync();
+                _exit(EXIT_SUCCESS);
+        }
+
+        log_info("Syncing filesystems and block devices");
```

Comment:
should probably end in a full stop

Reply:
Sorry, I don't quite follow. Can you clarify what is needed here?